From 16a285d307787bb9c3e2c5a27e228da5e035b3fe Mon Sep 17 00:00:00 2001
From: Bartosz Bilas <b.bilas@grinn-global.com>
Date: Thu, 2 Oct 2025 12:29:34 +0200
Subject: [PATCH] arch: mediatek: adjust memory size at runtime

Use TFA + libdram to obtain and set the actual memory size.
---
 arch/arm/include/asm/arch-mediatek/misc.h |  9 +++++++++
 arch/arm/mach-mediatek/mt8188/init.c      | 17 ++++++++++++-----
 2 files changed, 21 insertions(+), 5 deletions(-)

diff --git a/arch/arm/include/asm/arch-mediatek/misc.h b/arch/arm/include/asm/arch-mediatek/misc.h
index dd64c1f6..df1b9572 100644
--- a/arch/arm/include/asm/arch-mediatek/misc.h
+++ b/arch/arm/include/asm/arch-mediatek/misc.h
@@ -14,6 +14,15 @@
 #define APHW_VER		0x08
 #define APSW_VER		0x0c
 
+#define BOOT_ARGUMENT_LOCATION	(0x40000100)
+#define BOOT_ARGUMENT_MAGIC	0x504c504c
+#define BOOT_ARGUMENT		((struct boot_argument *)BOOT_ARGUMENT_LOCATION)
+
+struct boot_argument {
+	unsigned int magic_number;
+	unsigned long long dram_size;
+};
+
 void mediatek_capsule_update_board_setup(void);
 
 #endif /* __MEDIATEK_MISC_H_ */
diff --git a/arch/arm/mach-mediatek/mt8188/init.c b/arch/arm/mach-mediatek/mt8188/init.c
index 7d357f10..cf6dc1f6 100644
--- a/arch/arm/mach-mediatek/mt8188/init.c
+++ b/arch/arm/mach-mediatek/mt8188/init.c
@@ -21,11 +21,19 @@ int dram_init(void)
 {
 	int ret;
 
-	ret = fdtdec_setup_memory_banksize();
-	if (ret)
-		return ret;
+	if (BOOT_ARGUMENT->magic_number == BOOT_ARGUMENT_MAGIC) {
+		gd->ram_size = BOOT_ARGUMENT->dram_size;
+		gd->ram_base = 0x40000000UL;
+		debug("Boot argument DRAM size: %lluGB\n", gd->ram_size >> 30);
+	} else {
+		ret = fdtdec_setup_memory_banksize();
+		if (ret)
+			return ret;
 
-	fdtdec_setup_mem_size_base();
+		fdtdec_setup_mem_size_base();
+	}
+
+	mem_map[0].size = gd->ram_size;
 
 	/*
 	 * Limit gd->ram_top not exceeding SZ_4G.
@@ -82,7 +90,6 @@ static struct mm_region mt8188_mem_map[] = {
 		/* DDR */
 		.virt = 0x40000000UL,
 		.phys = 0x40000000UL,
-		.size = 0x200000000UL,
 		.attrs = PTE_BLOCK_MEMTYPE(MT_NORMAL) | PTE_BLOCK_OUTER_SHARE,
 	}, {
 		.virt = 0x00000000UL,
-- 
2.39.5
